readme.md
# Random StockPicker Cli
> A CLI tool that returns a random stock from a specific index

## Usage
```
NAME:
   randomsp-cli - a cli tool to randomly query stocks from a few major indices

USAGE:
   randomsp-cli [global options] command [command options] [arguments...]

VERSION:
   1.0.0

AUTHOR:
   Tyler Boright <ru.lai.development@gmail.com>

COMMANDS:
     random, rs  Return a random stock from any of the supported indices
     index, idx  Return a random stock from a specific index.  Defaults to S&P500.
     help, h     Shows a list of commands or help for one command

GLOBAL OPTIONS:
   --help, -h     show help
   --version, -v  print the version
```

## License
MIT © Tyler Boright


changelog.md
# RandomSP-Cli

## [1.2.1] - 2020-06-10
- Bumps version of dependency (randomsp) to fix bug where s&p 500 stocks were not being returned

## [1.2.0] - 2020-06-06 
- Adds support for grabbing nikkei 225 stocks
- Updates the Makefile to more clearly define the binary name when compiling


cli_test.go
package main

import (
	"fmt"
	"testing"
)

func TestStartCli(t *testing.T) {
	ran, err := StartCli([]string{"./rsp", "random"})
	if len(ran) < 1 && err != nil {
		fmt.Println("The random method was not able to find a stock.  You might not be connected to the internet.  Please connect and try again")
		t.Fail()
	}

	nas, err := StartCli([]string{"./rsp", "index", "nas"})
	if len(nas) < 1 && err != nil {
		fmt.Println("The nasdaq index method was unable to find a stock.  You might not be connected to the internet.  Please connect and try again")
		t.Fail()
	}

	sp, err := StartCli([]string{"./rsp", "index", "sp"})
	if len(sp) < 1 && err != nil {
		fmt.Println("The sp index method was unable to find a stock.  You might not be connected to the internet.  Please connect and try again")
		t.Fail()
	}

	dax, err := StartCli([]string{"./rsp", "index", "dax"})
	if len(dax) < 1 && err != nil {
		fmt.Println("The dax index method was unable to find a stock.  You might not be connected to the internet.  Please connect and try again")
		t.Fail()
	}

	ft, err := StartCli([]string{"./rsp", "index", "ft"})
	if len(ft) < 1 && err != nil {
		fmt.Println("The ft index method was unable to find a stock.  You might not be connected to the internet.  Please connect and try again")
		t.Fail()
	}

	ift, err := StartCli([]string{"./rsp", "index", "ift"})
	if len(ift) < 1 && err != nil {
		fmt.Println("The ift index method was unable to find a stock.  You might not be connected to the internet.  Please connect and try again")
		t.Fail()
	}
}


cli.go
package main

import (
	"fmt"
	rsp "github.com/tlboright/randomsp"
	"gopkg.in/urfave/cli.v1"
)

func StartCli(args []string) ([]string, error) {
	var resp []string
	app := cli.NewApp()
	app.Name = "randomsp-cli"
	app.Version = "1.1.0"
	app.Usage = "a cli tool to randomly query stocks from a few major indices"
	app.Authors = []cli.Author{
		cli.Author{
			Name:  "Tyler Boright",
			Email: "ru.lai.development@gmail.com",
		},
	}

	app.Action = func(c *cli.Context) error {
		resp = append(resp, "rsp: A stock index wasn't chosen\nFor help, type: rsp help\n")
		return nil
	}

	app.Commands = []cli.Command{
		{
			Name:    "random",
			Aliases: []string{"rs"},
			Usage:   "Return a random stock from any of the supported indices",
			Action: func(c *cli.Context) error {
				stock, err := rsp.GetRandomIndexStock()
				if err != nil {
					return fmt.Errorf("randomsp-cli failed with following error: %s. \nCheck your internet connection and try again.\n", err)
				}

				resp = append(resp, fmt.Sprintf("Your randomly selected stock symbol is:\n%s\n\n", stock.Symbol))
				resp = append(resp, fmt.Sprintf("%s is a stock on the %s stock index\n", stock.Symbol, stock.Index))

				return nil
			},
		},

		{
			Name:    "index",
			Aliases: []string{"idx"},
			Usage:   "Return a random stock from a specific index.  Defaults to S&P500.",
			Subcommands: []cli.Command{
				{
					Name:    "SP",
					Aliases: []string{"sp"},
					Usage:   "Returns a random stock from the Standard & Poors 500 index",
					Action: func(c *cli.Context) error {
						stock, err := rsp.GetRandomSPStock()
						if err != nil {
							return fmt.Errorf("randomsp-cli failed with following error: %s. \nCheck your internet connection and try again.\n", err)
						}

						resp = append(resp, fmt.Sprintf("Your randomly selected %s stock symbol is:\n%s\n", stock.Index, stock.Symbol))

						return nil
					},
				},
				{
					Name:    "DAX",
					Aliases: []string{"dax"},
					Usage:   "Returns a random stock from the DAX index",
					Action: func(c *cli.Context) error {
						stock, err := rsp.GetRandomDaxStock()
						if err != nil {
							return fmt.Errorf("randomsp-cli failed with following error: %s. \nCheck your internet connection and try again.\n", err)
						}

						resp = append(resp, fmt.Sprintf("Your randomly selected %s stock symbol is:\n%s\n", stock.Index, stock.Symbol))

						return nil
					},
				},
				{
					Name:    "FinancialTimes",
					Aliases: []string{"ft"},
					Usage:   "Returns a random stock from the Financial Times index",
					Action: func(c *cli.Context) error {
						stock, err := rsp.GetRandomFinancialTimesStock()
						if err != nil {
							return fmt.Errorf("randomsp-cli failed with following error: %s. \nCheck your internet connection and try again.\n", err)
						}

						resp = append(resp, fmt.Sprintf("Your randomly selected %s stock symbol is:\n%s\n", stock.Index, stock.Symbol))

						return nil
					},
				},
				{
					Name:    "Nasdaq",
					Aliases: []string{"nas", "nasdaq"},
					Usage:   "Returns a random stock from the Nasdaq index",
					Action: func(c *cli.Context) error {
						stock, err := rsp.GetRandomNasdaqStock()
						if err != nil {
							return fmt.Errorf("randomsp-cli failed with following error: %s. \nCheck your internet connection and try again.\n", err)
						}

						resp = append(resp, fmt.Sprintf("Your randomly selected %s stock symbol is:\n%s\n", stock.Index, stock.Symbol))

						return nil
					},
				},
				{
					Name:    "Nikkei",
					Aliases: []string{"nik", "nikkei"},
					Usage:   "Returns a random stock from the Nikkei index.\n    Note: Since the Nikkei 225 index uses numbers for symbols, the tool returns the name of the company and its number from the Nikkei 225 index",
					Action: func(c *cli.Context) error {
						stock, err := rsp.GetRandomNikkeiStock()
						if err != nil {
							return fmt.Errorf("randomsp-cli failed with following error: %s. \nCheck your internet connection and try again.\n", err)
						}

						resp = append(resp, fmt.Sprintf("Your randomly selected %s stock symbol is:\n%s\n", stock.Index, stock.Symbol))

						return nil
					},
				},
				{
					Name:    "ItalianFinancialTimes",
					Aliases: []string{"ift"},
					Usage:   "Returns a random stock from the Italian Financial Times index",
					Action: func(c *cli.Context) error {
						stock, err := rsp.GetRandomItalianFinancialTimesStock()
						if err != nil {
							return fmt.Errorf("randomsp-cli failed with following error: %s. \nCheck your internet connection and try again.\n", err)
						}

						resp = append(resp, fmt.Sprintf("Your randomly selected %s stock symbol is:\n%s\n", stock.Index, stock.Symbol))

						return nil
					},
				},
			},
		},
	}
	err := app.Run(args)
	return resp, err
}
